pennylane/transforms/metric_tensor.py

```diff
@@ -291,13 +387,247 @@ def processing_fn(probs):
         gs = []
 
         for prob, obs, coeffs in zip(probs, obs_list, coeffs_list):
+
             # calculate the covariance matrix of this layer
-            scale = qml.math.convert_like(np.outer(coeffs, coeffs), prob)
+            scale = qml.math.convert_like(qml.math.outer(coeffs, coeffs), prob)
             scale = qml.math.cast_like(scale, prob)
             g = scale * qml.math.cov_matrix(prob, obs, wires=tape.wires, diag_approx=diag_approx)
             gs.append(g)
 
         # create the block diagonal metric tensor
         return qml.math.block_diag(gs)
 
-    return metric_tensor_tapes, processing_fn
+    return metric_tensor_tapes, processing_fn, obs_list, coeffs_list
+
+
+@functools.lru_cache()
+def _get_gen_op(op, allow_nonunitary, aux_wire):
+    r"""Get the controlled-generator operation for a given operation.
+
+    Args:
+        op (pennylane.operation.Operation): Operation from which to extract the generator
+        allow_nonunitary (bool): Whether non-unitary gates are allowed in the circuit
+        aux_wire (int or pennylane.wires.Wires): Auxiliary wire on which to control the operation
+
+    Returns
+        qml.Operation: Controlled-generator operation of the generator of ``op``, controlled
+        on wire ``aux_wire``.
+
+    Raises
+        ValueError: If the generator of ``op`` is not known or it is non-unitary while
+        ``allow_nonunitary=False``.
+
+    If ``allow_nonunitary=True``, a general :class:`~.pennylane.ControlledQubitUnitary` is returned,
+    otherwise only controlled Pauli operations are used. If the operation has a non-unitary
+    generator but ``allow_nonunitary=False``, the operation ``op`` should have been decomposed
+    before, leading to a ``ValueError``.
+    """
+    gen, _ = op.generator
+    try:
+        if isinstance(gen, np.ndarray) or gen not in _GEN_TO_CGEN:
+            cgen = _OP_TO_CGEN[op.__class__]
+        else:
+            cgen = _GEN_TO_CGEN.get(gen, None)

```

Comment:
Minor, but the following might reduce the number of dict lookups by one:

```python
cgen = _GEN_TO_CGEN.get(gen, None)

if isinstance(gen, np.ndarray) or cgen is None:
    cgen = _OP_TO_CGEN[op.__class__]

Reply:
(if I am reading the logic correctly)

Reply:
Ah, this does not quite work, because if `gen` is an `ndarray` the first lookup will fail. I modified it slightly but still the saved lookup remains (although it is in a very small dictionary :D )